lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php

```diff
@@ -16,6 +16,83 @@
  * @access private
  */
 class WP_Theme_JSON_Resolver_6_1 extends WP_Theme_JSON_Resolver_6_0 {
+	/**
+	 * Returns the custom post type that contains the user's origin config
+	 * for the current theme or a void array if none are found.
+	 *
+	 * This can also create and return a new draft custom post type.
+	 *
+	 * @param WP_Theme $theme              The theme object.  If empty, it
+	 *                                     defaults to the current theme.
+	 * @param bool     $create_post        Optional. Whether a new custom post
+	 *                                     type should be created if none are
+	 *                                     found.  False by default.
+	 * @param array    $post_status_filter Filter Optional. custom post type by
+	 *                                     post status.  ['publish'] by default,
+	 *                                     so it only fetches published posts.
+	 * @return array Custom Post Type for the user's origin config.
+	 */
+	public static function get_user_data_from_wp_global_styles( $theme, $create_post = false, $post_status_filter = array( 'publish' ) ) {
+		if ( ! $theme instanceof WP_Theme ) {
+			$theme = wp_get_theme();
+		}
+		$user_cpt         = array();
+		$post_type_filter = 'wp_global_styles';
+		$args             = array(
+			'posts_per_page'      => 1,
+			'orderby'             => 'post_date',
+			'order'               => 'desc',
+			'post_type'           => $post_type_filter,
+			'post_status'         => $post_status_filter,
+			'tax_query'           => array(
+				array(
+					'taxonomy' => 'wp_theme',
+					'field'    => 'name',
+					'terms'    => $theme->get_stylesheet(),
+				),
+			),
+			'ignore_sticky_posts' => true,
+			'no_found_rows'       => true,
+		);
+
+		$cache_key = sprintf( 'wp_global_styles_%s', md5( serialize( $args ) ) );
+		$post_id   = (int) get_transient( $cache_key );
+		// Special case: '-1' is a results not found.
+		if ( -1 === $post_id && ! $create_post ) {
+			return $user_cpt;
+		}
+
+		if ( $post_id > 0 && in_array( get_post_status( $post_id ), (array) $post_status_filter, true ) ) {
+			return get_post( $post_id, ARRAY_A );
+		}
+
+		$gs_query     = new WP_Query();
+		$recent_posts = $gs_query->query( $args );
+		if ( count( $recent_posts ) === 1 ) {
+			$user_cpt = get_post( $recent_posts[0], ARRAY_A );
+		} elseif ( $create_post ) {
+			$cpt_post_id = wp_insert_post(
+				array(
+					'post_content' => '{"version": ' . WP_Theme_JSON_Gutenberg::LATEST_SCHEMA . ', "isGlobalStylesUserThemeJSON": true }',
+					'post_status'  => 'publish',
+					'post_title'   => __( 'Custom Styles', 'default' ),
+					'post_type'    => $post_type_filter,
+					'post_name'    => 'wp-global-styles-' . urlencode( wp_get_theme()->get_stylesheet() ),
```

Comment:
The addition of `sprintf` from https://github.com/WordPress/wordpress-develop/pull/2414/files#diff-d6b86476eed058e7cf8b6e57fa52c4fd75b1f0907e1a9ccb0149528a24f7578bR298 seems legit. Shouldn't this be a copy of that other PR?